common/persistence/versionHistory.go

```diff
@@ -21,174 +21,517 @@
-				currentItem.Version,
-				lastItem.Version),
+	for _, item := range inputitems {
+		if err := versionHistory.AddOrUpdateItem(item.Duplicate()); err != nil {
+			panic(fmt.Sprintf("unable to initialize version history: %v", err))
 		}
 	}
 
-	if currentItem.EventID <= lastItem.EventID {
-		return &shared.BadRequestError{
-			Message: fmt.Sprintf("cannot add version history with a lower event id %v. Last event id: %v",
-				currentItem.EventID,
-				lastItem.EventID),
+	return versionHistory
+}
+
+// NewVersionHistoryFromThrift create a new version history from thrift object
+func NewVersionHistoryFromThrift(
+	input *shared.VersionHistory,
+) *VersionHistory {
+
+	if input == nil {
+		panic("version history is null")
+	}
+
+	items := []*VersionHistoryItem{}
+	for _, item := range input.Items {
+		items = append(items, NewVersionHistoryItemFromThrift(item))
+	}
+	return NewVersionHistory(input.BranchToken, items)
+}
+
+// Duplicate duplicate VersionHistory
+func (v *VersionHistory) Duplicate() *VersionHistory {
+
+	return NewVersionHistory(v.branchToken, v.items)
+}
+
+// ToThrift return thrift format of version history
+func (v *VersionHistory) ToThrift() *shared.VersionHistory {
+
+	token := make([]byte, len(v.branchToken))
+	copy(token, v.branchToken)
+	items := []*shared.VersionHistoryItem{}
+	for _, item := range v.items {
+		items = append(items, item.ToThrift())
+	}
+
+	tHistory := &shared.VersionHistory{
+		BranchToken: token,
+		Items:       items,
+	}
+	return tHistory
+}
+
+// DuplicateUntilLCAItem duplicate the version history up until LCA item
+func (v *VersionHistory) DuplicateUntilLCAItem(
+	lcaItem *VersionHistoryItem,
+) (*VersionHistory, error) {
+
+	versionHistory := NewVersionHistory(nil, nil)
+	notFoundErr := &shared.BadRequestError{
+		Message: "version history does not contains the LCA item.",
+	}
+	for _, item := range v.items {
+
+		if item.version < lcaItem.version {
+			if err := versionHistory.AddOrUpdateItem(item); err != nil {
+				return nil, err
+			}
+
+		} else if item.version == lcaItem.version {
+			if lcaItem.eventID > item.eventID {
+				return nil, notFoundErr
+			}
+			if err := versionHistory.AddOrUpdateItem(lcaItem); err != nil {
+				return nil, err
+			}
+			return versionHistory, nil
+
+		} else {
+			return nil, notFoundErr
 		}
 	}
 
-	if currentItem.Version > lastItem.Version {
+	return nil, notFoundErr
+}
+
+// SetBranchToken the set the branch token
+func (v *VersionHistory) SetBranchToken(
+	inputToken []byte,
+) error {
+
+	if len(v.branchToken) != 0 {
+		return &shared.BadRequestError{Message: "cannot branch token is already set"}
+	}
+
+	token := make([]byte, len(inputToken))
+	copy(token, inputToken)
+	v.branchToken = token
+	return nil
+}
+
+// GetBranchToken return the branch token
+func (v *VersionHistory) GetBranchToken() []byte {
+	token := make([]byte, len(v.branchToken))
+	copy(token, v.branchToken)
```

Comment:
Why copying this everywhere?

Reply:
i want to make this immutable, to achieve less bug, perf is not a concern, for now